### PR TITLE
Updated Q# target tests

### PR DESCRIPTION
**Pull request by troelsfr**
This PR introduces QAT validation of Q# target packages into the CI pipeline. This PR contains:

- Streamlining of provider names to match target names used in Q# (with backwards compatibility)
- Tests of around 100 operations for each of the targets
- Github CI integration
- Testing of some basic Q# codeblocks for each of the targets

### CODE REVIEW

File: qir/qat/ValidationPass/ValidationPassConfiguration.cpp [hunk 1]
```diff
@@ -79,7 +79,7 @@ namespace quantum
             profile.allowed_pointer_types_   = {"i8*",    "i16*",    "i32*",    "i64*",
                                               "Qubit*", "Qubit**", "Result*", "Result**"};
         }
-        else if (name == "provider_4bf9")
+        else if (name == "provider_4bf9" || name == "provider_3")
```
Comment (bettinaheim) [inline qir/qat/ValidationPass/ValidationPassConfiguration.cpp:82]: We said to remove this change.

File: qir/qsharp-target-tests/TargetTestScripts/__init__.py [hunk 2]
```diff
@@ -1,63 +1,509 @@
...
+    "Decompositions/CYAdjoint.qs",
+    "Decompositions/CYControlled.qs",
+    "Decompositions/CYAdjointControlled.qs",
+    "Decompositions/CYControlledAdjoint.qs",
+    "Decompositions/CZ.qs",
+    "Decompositions/CZAdjoint.qs",
+    "Decompositions/CZControlled.qs",
+    "Decompositions/CZAdjointControlled.qs",
+    "Decompositions/CZControlledAdjoint.qs",
+    # TODO(tfr): "Decompositions/ExpFrac.qs",
```
Comment (bettinaheim) [inline qir/qsharp-target-tests/TargetTestScripts/__init__.py:37]: How come these are a todo? They are not supported in any of the profiles, so I don't think there should be tests for them.

File: qir/qsharp-target-tests/TargetTestScripts/__init__.py [hunk 3]
```diff
@@ -1,63 +1,509 @@
...
+    "Decompositions/Z.qs",
+    "Decompositions/ZAdjoint.qs",
+    "Decompositions/ZControlled.qs",
+    "Decompositions/ZAdjointControlled.qs",
+    "Decompositions/ZControlledAdjoint.qs",
+
+
+    "CodeBlocks/QubitArrays1.qs",
+    "CodeBlocks/QubitArrays2.qs",
+    # TODO(tfr): Does not compile "CodeBlocks/Range1.qs",
```
Comment (bettinaheim) [inline qir/qsharp-target-tests/TargetTestScripts/__init__.py:131]: What issue is this one?

File: qir/qsharp-target-tests/TargetTestScripts/__init__.py [hunk 4]
```diff
@@ -1,63 +1,509 @@
...
+    "Decompositions/HControlled.qs",
+    "Decompositions/HAdjointControlled.qs",
+    "Decompositions/HControlledAdjoint.qs",
+    "Decompositions/I.qs",
+    "Decompositions/IAdjoint.qs",
+    "Decompositions/IControlled.qs",
+    "Decompositions/IAdjointControlled.qs",
+    "Decompositions/IControlledAdjoint.qs",
+    "Decompositions/M.qs",
+    # TODO(tfr): not straightline input "Decompositions/MResetX.qs",
```
Comment (bettinaheim) [inline qir/qsharp-target-tests/TargetTestScripts/__init__.py:305]: Is not straightline input addressed in a subsequent PR?

File: qir/qsharp-target-tests/TargetTestScripts/__init__.py [hunk 5]
```diff
@@ -1,63 +1,509 @@
...
+    "Decompositions/SAdjoint.qs",
+    "Decompositions/SControlled.qs",
+    "Decompositions/SAdjointControlled.qs",
+    "Decompositions/SControlledAdjoint.qs",
+    "Decompositions/SWAP.qs",
+    "Decompositions/SWAPAdjoint.qs",
+    "Decompositions/SWAPControlled.qs",
+    "Decompositions/SWAPAdjointControlled.qs",
+    "Decompositions/SWAPControlledAdjoint.qs",
+    # TODO(tfr): Not supported by hardware "Decompositions/SetToBasisState.qs",
```
Comment (bettinaheim) [inline qir/qsharp-target-tests/TargetTestScripts/__init__.py:358]: Will be resolved after a package update.

File: qir/qsharp-target-tests/TargetTestScripts/CodeBlocks/broke.NestedArrays.qs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    open Microsoft.Quantum.Intrinsic;
+
+    @EntryPoint()
+    operation Main(): Result
+    {
+
+        use qubits = Qubit[10];
+        let xx = [[[[[[[[[[[[ [qubits, qubits] ]]]]]]]]]]]];
+        
+        CNOT(xx[0][0][0][0][0][0][0][0][0][0][0][0][Length(xx[0][0][0][0][0][0][0][0][0][0][0][0]) -1][Length(xx[0][0][0][0][0][0][0][0][0][0][0][0][Length(xx[0][0][0][0][0][0][0][0][0][0][0][0]) - 1]) - 1], q);
```
[the remark this answers is not in the document]
Comment (troelsfr) [reply qir/qsharp-target-tests/TargetTestScripts/CodeBlocks/broke.NestedArrays.qs:14]: No, these need to be deleted again.

File: qir/qsharp-target-tests/TargetTestScripts/__init__.py [hunk 7]
```diff
@@ -1,63 +1,509 @@
...
+    "Decompositions/ZControlled.qs",
+    "Decompositions/ZAdjointControlled.qs",
+    "Decompositions/ZControlledAdjoint.qs",
+
+
+    "CodeBlocks/QubitArrays1.qs",
+    "CodeBlocks/QubitArrays2.qs",
+    # TODO(tfr): Does not compile "CodeBlocks/Range1.qs",
+    "CodeBlocks/OffsetBySix.qs",
+    # TODO(tfr): Does not compile "CodeBlocks/ConstantCircuitUsingMutables1.qs",
```
Comment (bettinaheim) [inline qir/qsharp-target-tests/TargetTestScripts/__init__.py:385]: What does "does not compile" mean; is the Q# compiler giving a nice error stating it is not supported on that target?

File: qir/qsharp-target-tests/TargetTestScripts/__init__.py (hunk 2 above)
Comment (troelsfr) [reply qir/qsharp-target-tests/TargetTestScripts/__init__.py:37]: OK, I will remove them.

File: qir/qsharp-target-tests/TargetTestScripts/__init__.py (hunk 4 above)
Comment (troelsfr) [reply qir/qsharp-target-tests/TargetTestScripts/__init__.py:305]: Yes, it is addressed in another PR.

File: qir/qsharp-target-tests/TargetTestScripts/__init__.py (hunk 7 above)
Comment (troelsfr) [reply qir/qsharp-target-tests/TargetTestScripts/__init__.py:385]: It can mean either. It is simply scripts that I would have expected to be deployable that for one reason or the other does not make it through Q#

File: qir/qsharp-target-tests/template/Example.csproj [hunk 8]
```diff
@@ -1,19 +1,10 @@
...
     <TargetFramework>net6.0</TargetFramework>
-    <QirGeneration>true</QirGeneration>
-    <AdditionalQscArguments>--runtime-override AdaptiveExecution</AdditionalQscArguments>
-    <IncludeCSharpRuntime>False</IncludeCSharpRuntime>
-    <IncludeQSharpCorePackages>False</IncludeQSharpCorePackages>
     <IncludeProviderPackages>False</IncludeProviderPackages>
-    <CSharpGeneration>False</CSharpGeneration>
-
+    <ExecutionTarget>{{executionTarget}}</ExecutionTarget>
+    <QirGeneration>true</QirGeneration>
```
Comment (bettinaheim) [inline qir/qsharp-target-tests/template/Example.csproj:7]: This isn't needed unless you want the human readable .ll output. The bitcode is produced without it and can be found in the obj/qsharp folder of the project. 
```suggestion
```

File: qir/qsharp/TargetPackages/QSharpVersion/Example.csproj [hunk 9]
```diff
@@ -1,19 +1,13 @@
...
-    <AdditionalQscArguments>--runtime-override AdaptiveExecution</AdditionalQscArguments>
-    <IncludeCSharpRuntime>False</IncludeCSharpRuntime>
-    <IncludeQSharpCorePackages>False</IncludeQSharpCorePackages>
     <IncludeProviderPackages>False</IncludeProviderPackages>
-    <CSharpGeneration>False</CSharpGeneration>
-
+    <ExecutionTarget>qci.qpu</ExecutionTarget>
+    <!--ExecutionTarget>rigetti.qpu</ExecutionTarget-->
+    <!--TargetCapability>AdaptiveExecution</TargetCapability-->
+    <QirGeneration>true</QirGeneration>
```
Comment (bettinaheim) [inline qir/qsharp/TargetPackages/QSharpVersion/Example.csproj:10]: Same here. This is not needed.

File: qir/qsharp-target-tests/TargetTestScripts/__init__.py (hunk 7 above)
Comment (bettinaheim) [reply qir/qsharp-target-tests/TargetTestScripts/__init__.py:385]: Ok, we need to go over these and make a call on what to do with each one - if the error is justified, then I think we should remove them here. We can proceed with this PR without that, but could you please file a GitHub issue to go over this and either remove or enable the test cases?

File: qir/qsharp-target-tests/TargetTestScripts/__init__.py (hunk 3 above)
Comment (troelsfr) [reply qir/qsharp-target-tests/TargetTestScripts/__init__.py:131]: The issue is slicing using `...` - it is not really a qat problem, so I remove this test.

File: qir/qsharp-target-tests/template/Example.csproj (hunk 8 above)
Comment (troelsfr) [reply qir/qsharp-target-tests/template/Example.csproj:7]: It is needed for test output

File: qir/qsharp/TargetPackages/QSharpVersion/Example.csproj (hunk 9 above)
Comment (troelsfr) [reply qir/qsharp/TargetPackages/QSharpVersion/Example.csproj]: I use this project for developing tests, so it is kind of handy to have here.

**Review by bettinaheim — APPROVED** [review]

File: qir/qat/ValidationPass/ValidationPassConfiguration.cpp [hunk 10]
```diff
@@ -79,7 +79,7 @@ namespace quantum
             profile.allowed_pointer_types_   = {"i8*",    "i16*",    "i32*",    "i64*",
                                               "Qubit*", "Qubit**", "Result*", "Result**"};
         }
-        else if (name == "provider_4bf9")
+        else if (name == "provider_b340bf9")
```
Comment (bettinaheim) [inline qir/qat/ValidationPass/ValidationPassConfiguration.cpp:82]: @troelsfr Please undo this change.
